File: runtime/node/vssh/fila.js
'use strict';

// A fila de processamento: delegar um container ao cluster Kubernetes. O par de `vssh/fila.py`.
//
// O app declara `recursos.fila: true` no manifesto e recebe do portal, a cada subida,
// `VSSH_PORTAL_URL` e `VSSH_PORTAL_TOKEN`. É com esse par que tudo aqui fala com `/api/fila/*`.
// Sem ele, `disponivel()` diz o motivo e `submeter()` rejeita com `ErroDaFila`.
//
//   const { fila } = require('vssh');
//   const id = await fila.submeter({ imagem: 'ghcr.io/colabhd/ffmpeg:7', comando: ['ffmpeg', ...],
//     entradas: { 'v.mp4': '/home/ana/v.mp4' }, saidas: ['v.webm'], gpu: { quantidade: 1 } });
//   const final = await fila.acompanhar(id, (evento, job) => console.log(evento, job.estado));
//   if (final.estado === 'concluido') await fila.baixar(id, '/home/ana/saida');
//
// Estados: `declarado`, `enviado`, `na_fila`, `rodando`, e os finais `concluido`, `falhou`
// (`motivo`: `prazo`, `imagem`, `entrada:<nome>`, `codigo:<n>`, `sumiu`) e `cancelado`.
//
// Um trabalho que imprime linhas de `vssh/progresso` no stdout tem o progresso lido pelo portal:
// o `aoEvento` de `acompanhar` recebe `progresso` com o job, e `job.progresso` traz `feito`,
// `total` e `etapa`. Quem já baixou as saídas chama `remover` para o dado sair do S3 na hora.

const fs = require('node:fs');
const path = require('node:path');
const http = require('node:http');
const https = require('node:https');

class ErroDaFila extends Error {
  constructor(status, mensagem, extra = {}) { super(mensagem); this.status = status; this.mensagem = mensagem; this.extra = extra; }
}

function credencial(env = process.env) {
  const url = (env.VSSH_PORTAL_URL || '').replace(/\/+$/, '');
  const token = env.VSSH_PORTAL_TOKEN || '';
  return url && token ? { url, token } : null;
}

function modulo(url) { return url.startsWith('https:') ? https : http; }

/** Um pedido ao portal. Com `stream`, resolve com a resposta crua (para SSE). */
// O SDK se apresenta pelo nome, como o de Python: o portal fica atrás do Cloudflare, e a regra de
// bots dele julga pelo User-Agent. O Node não manda nenhum, e passa; o nome diz no log quem chamou.
const AGENTE = 'vssh-sdk-fila/node';

function pedir(metodo, rota, { corpo, env, stream = false } = {}) {
  const cred = credencial(env);
  if (!cred) return Promise.reject(new ErroDaFila(0, 'o app não tem credencial da fila: declare recursos.fila no manifesto e reinicie o app'));
  const dados = corpo === undefined ? null : Buffer.from(JSON.stringify(corpo), 'utf8');
  const u = new URL(cred.url + '/api/fila' + rota);
  return new Promise((resolve, reject) => {
    const req = modulo(u.href).request(u, {
      method: metodo,
      headers: {
        authorization: `Bearer ${cred.token}`, accept: 'application/json', 'user-agent': AGENTE,
        ...(dados ? { 'content-type': 'application/json', 'content-length': String(dados.length) } : {}),
      },
      timeout: stream ? 0 : 60_000,
    }, (res) => {
      if (stream && res.statusCode === 200) { resolve(res); return; }
      const pedacos = [];
      res.on('data', (d) => pedacos.push(d));
      res.on('end', () => {
        const texto = Buffer.concat(pedacos).toString('utf8');
        let j = null; try { j = texto ? JSON.parse(texto) : null; } catch { /* texto cru */ }
        if (res.statusCode >= 200 && res.statusCode < 300) { resolve(j); return; }
        const { error, ...extra } = j || {};
        reject(new ErroDaFila(res.statusCode, error || texto || `HTTP ${res.statusCode}`, extra));
      });
    });
    req.on('timeout', () => req.destroy(new Error('o portal não respondeu a tempo')));
    req.on('error', (e) => reject(new ErroDaFila(0, `o portal não respondeu: ${e.message}`)));
    if (dados) req.write(dados);
    req.end();
  });
}

/** Nunca rejeita: sem credencial, ou com o portal fora, responde `disponivel: false` com o motivo. */
async function disponivel(env = process.env) {
  const vazio = { cluster: null, gpus: [], quotas: {}, uso: {} };
  if (!credencial(env)) return { disponivel: false, motivo: 'o app não declarou recursos.fila no manifesto (ou ainda não foi reiniciado)', ...vazio };
  try {
    const r = await pedir('GET', '', { env });
    return { disponivel: !!r.disponivel, motivo: r.motivo ?? null, cluster: r.cluster ?? null, gpus: r.gpus || [], quotas: r.quotas || {}, uso: r.uso || {} };
  } catch (e) {
    return { disponivel: false, motivo: e.mensagem || e.message, ...vazio };
  }
}

function entradasDe(trabalho) {
  const bruto = trabalho.entradas || {};
  const pares = Array.isArray(bruto) ? bruto.map((c) => [path.basename(c), c]) : Object.entries(bruto);
  const vistos = new Set();
  for (const [nome, caminho] of pares) {
    if (vistos.has(nome)) throw new ErroDaFila(400, `duas entradas com o mesmo nome: ${nome}`);
    vistos.add(nome);
    if (!fs.existsSync(caminho) || !fs.statSync(caminho).isFile()) throw new ErroDaFila(400, `entrada não encontrada: ${caminho}`);
  }
  return pares;
}

function subir(url, caminho) {
  const tamanho = fs.statSync(caminho).size;
  const u = new URL(url);
  return new Promise((resolve, reject) => {
    const req = modulo(u.href).request(u, { method: 'PUT', headers: { 'content-length': String(tamanho), 'content-type': 'application/octet-stream' } }, (res) => {
      res.resume();
      res.on('end', () => (res.statusCode >= 200 && res.statusCode < 300)
        ? resolve(tamanho)
        : reject(new ErroDaFila(res.statusCode, `o S3 recusou a entrada ${caminho} (${res.statusCode})`)));
    });
    req.on('error', (e) => reject(new ErroDaFila(0, `o S3 não respondeu ao subir ${caminho}: ${e.message}`)));
    fs.createReadStream(caminho).on('error', reject).pipe(req);
  });
}

/**
 * Declara, sobe as entradas e inicia. Resolve com o id do job. `trabalho` leva `imagem`,
 * `comando`, `args`, `env`, `entradas` (objeto nome → caminho, ou lista de caminhos), `saidas`,
 * `gpu`, `cpu`, `memoria`, `disco`, `prazo` e `nome`.
 */
async function submeter(trabalho, aoProgresso, env = process.env) {
  const pares = entradasDe(trabalho);
  const { entradas: _ignorado, ...pedido } = trabalho;
  pedido.entradas = pares.map(([nome]) => nome);
  const decl = await pedir('POST', '/jobs', { corpo: pedido, env });
  const urls = Object.fromEntries((decl.entradas || []).map((e) => [e.nome, e.url]));
  try {
    for (const [nome, caminho] of pares) {
      if (decl.maxEntradaBytes && fs.statSync(caminho).size > decl.maxEntradaBytes) {
        throw new ErroDaFila(413, `a entrada ${caminho} passa do teto de ${decl.maxEntradaBytes} bytes`);
      }
      const bytes = await subir(urls[nome], caminho);
      if (aoProgresso) { try { aoProgresso({ fase: 'entrada', arquivo: caminho, bytes }); } catch { /* do app */ } }
    }
    await pedir('POST', `/jobs/${decl.id}/iniciar`, { env });
  } catch (e) {
    await pedir('DELETE', `/jobs/${decl.id}`, { env }).catch(() => {});
    throw e;
  }
  return decl.id;
}

const estado = (id, env = process.env) => pedir('GET', `/jobs/${id}`, { env });

/** Itera `{ evento, dados }` de um SSE do portal até ele fechar. */
async function* eventos(rota, env) {
  const res = await pedir('GET', rota, { env, stream: true });
  let resto = '';
  for await (const pedaco of res) {
    resto += pedaco.toString('utf8');
    const blocos = resto.split('\n\n');
    resto = blocos.pop();
    for (const b of blocos) {
      const evento = /^event: (.+)$/m.exec(b)?.[1];
      const dados = /^data: (.+)$/m.exec(b)?.[1];
      if (!evento) continue;
      let j = null; try { j = dados ? JSON.parse(dados) : null; } catch { /* sem dados */ }
      yield { evento, dados: j };
    }
  }
}

const FINAIS = new Set(['concluido', 'falhou', 'cancelado']);

/** Segue o job até um estado final e resolve com ele. `aoEvento(evento, job)` a cada mudança. */
async function acompanhar(id, aoEvento, env = process.env) {
  for (;;) {
    let final = null;
    for await (const { evento, dados } of eventos(`/jobs/${id}/eventos`, env)) {
      if (aoEvento && dados) { try { aoEvento(evento, dados); } catch { /* do app */ } }
      if (FINAIS.has(evento)) final = dados;
    }
    if (final) return final;
    const atual = await estado(id, env);
    if (FINAIS.has(atual.estado)) return atual;
  }
}

/** As linhas do container principal, como iterador assíncrono. */
async function* log(id, env = process.env) {
  for await (const { evento, dados } of eventos(`/jobs/${id}/log`, env)) {
    if (evento === 'linha' && dados) yield dados.linha || '';
    else if (evento === 'error' && dados) throw new ErroDaFila(0, dados.error || 'erro no log');
  }
}

async function cancelar(id, env = process.env) {
  const r = await pedir('POST', `/jobs/${id}/cancelar`, { env });
  return !!(r && r.success);
}

/** Baixa as saídas para `destino` e resolve com os caminhos gravados. */
async function baixar(id, destino, nomes, env = process.env) {
  const r = await pedir('GET', `/jobs/${id}/saidas`, { env });
  fs.mkdirSync(destino, { recursive: true });
  const gravados = [];
  for (const s of r.saidas || []) {
    if (nomes && !nomes.includes(s.nome)) continue;
    const caminho = path.join(destino, s.nome);
    await new Promise((resolve, reject) => {
      const u = new URL(s.url);
      modulo(u.href).get(u, (res) => {
        if (res.statusCode !== 200) { res.resume(); reject(new ErroDaFila(res.statusCode, `o S3 recusou a saída ${s.nome} (${res.statusCode})`)); return; }
        const f = fs.createWriteStream(caminho);
        res.pipe(f);
        f.on('finish', resolve);
        f.on('error', reject);
      }).on('error', (e) => reject(new ErroDaFila(0, `o S3 não respondeu ao baixar ${s.nome}: ${e.message}`)));
    });
    gravados.push(caminho);
  }
  return gravados;
}

async function listar(env = process.env) {
  const r = await pedir('GET', '/jobs', { env });
  return (r && r.jobs) || [];
}

/**
 * Apaga do S3 as entradas e as saídas de um job terminado, sem esperar a faxina da retenção. Serve
 * a quem já baixou o que precisava e não quer o dado no cluster nem mais um minuto. Um job em
 * curso responde 409: cancele antes.
 */
async function remover(id, env = process.env) {
  const r = await pedir('DELETE', `/jobs/${id}`, { env });
  return !!(r && r.removido);
}

module.exports = { ErroDaFila, disponivel, submeter, estado, acompanhar, log, cancelar, remover, baixar, listar };

File: runtime/python/vssh/fila.py
"""
A fila de processamento: delegar um container ao cluster Kubernetes, em vez de rodá-lo aqui.

O app declara `recursos.fila: true` no manifesto e recebe do portal, a cada subida,
`VSSH_PORTAL_URL` e `VSSH_PORTAL_TOKEN`: a URL pública do portal e a identidade deste app diante
dele. É com esse par que tudo aqui fala com `/api/fila/*`. Sem ele, `disponivel()` diz o motivo e
`submeter()` levanta `ErroDaFila`; um servidor sem cluster configurado responde o mesmo, pela
mesma pergunta. O app pergunta antes de prometer um botão.

O trabalho é um container: a imagem do app, o comando, arquivos de entrada da estação e os nomes
das saídas que ele vai escrever em `/vssh/saidas`. Os arquivos viajam pelo S3 do cluster, por URLs
que o portal assina; este módulo sobe as entradas e baixa as saídas com `urllib`, em stream.

    from vssh import fila

    if fila.disponivel()['disponivel']:
        ident = fila.submeter({
            'nome': 'transcodificar',
            'imagem': 'ghcr.io/colabhd/ffmpeg:7',
            'comando': ['ffmpeg', '-i', 'entradas/v.mp4', 'saidas/v.webm'],
            'entradas': {'v.mp4': '/home/ana/videos/v.mp4'},
            'saidas': ['v.webm'],
            'gpu': {'quantidade': 1, 'tipo': 'rtx-a5000'},
            'cpu': '4', 'memoria': '8Gi', 'prazo': 3600,
        })
        final = fila.acompanhar(ident, ao_evento=lambda ev, job: print(ev, job['estado']))
        if final['estado'] == 'concluido':
            fila.baixar(ident, '/home/ana/videos/saida')

Estados de um job: `declarado`, `enviado`, `na_fila`, `rodando`, e os finais `concluido`,
`falhou` (com `motivo`: `prazo`, `imagem`, `entrada:<nome>`, `codigo:<n>`, `sumiu`) e
`cancelado`. Dentro do container, `VSSH_ENTRADAS` e `VSSH_SAIDAS` apontam os dois diretórios, e o
diretório de trabalho é `/vssh`.

Um trabalho que imprime linhas de `vssh.progresso` no stdout tem o progresso lido pelo portal: o
`ao_evento` de `acompanhar` recebe `progresso` com o job, e `job['progresso']` traz `feito`,
`total` e `etapa`. Quem já baixou as saídas chama `remover` para o dado sair do S3 na hora.

Só biblioteca padrão, como o resto do pacote.
"""

import json
import os
import urllib.error
import urllib.request

__all__ = ['ErroDaFila', 'disponivel', 'submeter', 'estado', 'acompanhar', 'log', 'cancelar', 'remover', 'baixar', 'listar']

BLOCO = 4 * 1024 * 1024
_TEMPO_HTTP = 60


class ErroDaFila(Exception):
    """Uma recusa do portal (`status`, `mensagem`) ou a ausência da credencial (`status` 0)."""

    def __init__(self, status, mensagem, extra=None):
        super().__init__(mensagem)
        self.status = status
        self.mensagem = mensagem
        self.extra = extra or {}


def _credencial(env=None):
    env = os.environ if env is None else env
    url = (env.get('VSSH_PORTAL_URL') or '').rstrip('/')
    token = env.get('VSSH_PORTAL_TOKEN') or ''
    if not url or not token:
        return None
    return url, token


# O SDK se apresenta pelo nome. O portal fica atrás do Cloudflare, e a regra de bots dele
# recusa o `Python-urllib/3.x` que o urllib manda por padrão (erro 1010, "browser signature
# banned"); um nome próprio passa, e diz no log do portal quem chamou.
_AGENTE = 'vssh-sdk-fila/python'


def _pedir(metodo, rota, corpo=None, env=None, stream=False):
    cred = _credencial(env)
    if not cred:
        raise ErroDaFila(0, 'o app não tem credencial da fila: declare recursos.fila no manifesto e reinicie o app')
    url, token = cred
    dados = None
    cabecalhos = {'Authorization': 'Bearer ' + token, 'Accept': 'application/json', 'User-Agent': _AGENTE}
    if corpo is not None:
        dados = json.dumps(corpo).encode('utf-8')
        cabecalhos['Content-Type'] = 'application/json'
    req = urllib.request.Request(url + '/api/fila' + rota, data=dados, method=metodo, headers=cabecalhos)
    try:
        r = urllib.request.urlopen(req, timeout=None if stream else _TEMPO_HTTP)
    except urllib.error.HTTPError as e:
        texto = e.read().decode('utf-8', 'replace')
        try:
            j = json.loads(texto)
        except ValueError:
            j = {}
        raise ErroDaFila(e.code, j.get('error') or texto or ('HTTP %d' % e.code), {k: v for k, v in j.items() if k != 'error'})
    except urllib.error.URLError as e:
        raise ErroDaFila(0, 'o portal não respondeu: %s' % e.reason)
    if stream:
        return r
    with r:
        texto = r.read().decode('utf-8')
    return json.loads(texto) if texto else None


def disponivel(env=None):
    """`{'disponivel': bool, 'motivo': str | None, 'cluster': str | None, 'gpus': [...], 'quotas': {...}, 'uso': {...}}`.

    Nunca levanta: sem credencial, ou com o portal fora, responde `disponivel: False` com o motivo.
    """
    if not _credencial(env):
        return {'disponivel': False, 'motivo': 'o app não declarou recursos.fila no manifesto (ou ainda não foi reiniciado)',
                'cluster': None, 'gpus': [], 'quotas': {}, 'uso': {}}
    try:
        r = _pedir('GET', '', env=env)
    except ErroDaFila as e:
        return {'disponivel': False, 'motivo': e.mensagem, 'cluster': None, 'gpus': [], 'quotas': {}, 'uso': {}}
    return {
        'disponivel': bool(r.get('disponivel')),
        'motivo': r.get('motivo'),
        'cluster': r.get('cluster'),
        'gpus': r.get('gpus') or [],
        'quotas': r.get('quotas') or {},
        'uso': r.get('uso') or {},
    }


def _entradas_de(trabalho):
    """`{'nome': caminho}` ou `[caminho, ...]` (nome = basename; colisão levanta)."""
    bruto = trabalho.get('entradas') or {}
    if isinstance(bruto, dict):
        pares = list(bruto.items())
    else:
        pares = [(os.path.basename(c), c) for c in bruto]
    vistos = set()
    for nome, caminho in pares:
        if nome in vistos:
            raise ErroDaFila(400, 'duas entradas com o mesmo nome: %s' % nome)
        vistos.add(nome)
        if not os.path.isfile(caminho):
            raise ErroDaFila(400, 'entrada não encontrada: %s' % caminho)
    return pares


def _subir(url, caminho, ao_progresso=None):
    tamanho = os.path.getsize(caminho)
    with open(caminho, 'rb') as f:
        req = urllib.request.Request(url, data=f, method='PUT',
                                     headers={'Content-Length': str(tamanho), 'Content-Type': 'application/octet-stream'})
        try:
            with urllib.request.urlopen(req, timeout=None) as r:
                if r.status not in (200, 201, 204):
                    raise ErroDaFila(r.status, 'o S3 recusou a entrada %s' % caminho)
        except urllib.error.HTTPError as e:
            raise ErroDaFila(e.code, 'o S3 recusou a entrada %s (%d)' % (caminho, e.code))
        except urllib.error.URLError as e:
            raise ErroDaFila(0, 'o S3 não respondeu ao subir %s: %s' % (caminho, e.reason))
    if ao_progresso:
        ao_progresso({'fase': 'entrada', 'arquivo': caminho, 'bytes': tamanho})


def submeter(trabalho, ao_progresso=None, env=None):
    """Declara, sobe as entradas e inicia. Devolve o id do job.

    `trabalho` leva `imagem` (obrigatória), `comando`, `args`, `env`, `entradas` (dict nome →
    caminho, ou lista de caminhos), `saidas` (nomes em /vssh/saidas), `gpu` (`{'quantidade',
    'tipo'}`), `cpu`, `memoria`, `disco`, `prazo` (segundos) e `nome`. O que o portal recusa vem
    como `ErroDaFila` com a mensagem dele, antes de qualquer byte subir.
    """
    pares = _entradas_de(trabalho)
    pedido = {k: v for k, v in trabalho.items() if k != 'entradas'}
    pedido['entradas'] = [nome for nome, _ in pares]
    decl = _pedir('POST', '/jobs', pedido, env=env)
    ident = decl['id']
    urls = {e['nome']: e['url'] for e in decl.get('entradas', [])}
    teto = decl.get('maxEntradaBytes')
    try:
        for nome, caminho in pares:
            if teto and os.path.getsize(caminho) > teto:
                raise ErroDaFila(413, 'a entrada %s passa do teto de %d bytes' % (caminho, teto))
            _subir(urls[nome], caminho, ao_progresso)
        _pedir('POST', '/jobs/%s/iniciar' % ident, env=env)
    except ErroDaFila:
        # O que ficou declarado não pode contar na cota de quem nem chegou a submeter.
        try:
            _pedir('DELETE', '/jobs/%s' % ident, env=env)
        except ErroDaFila:
            pass
        raise
    return ident


def estado(ident, env=None):
    """O job como o portal o vê: `id`, `estado`, `motivo`, `saidas`, `faltantes`, datas."""
    return _pedir('GET', '/jobs/%s' % ident, env=env)


def _eventos(rota, env=None):
    """Itera `(evento, dados)` de um SSE do portal até ele fechar."""
    r = _pedir('GET', rota, env=env, stream=True)
    with r:
        evento, linhas = None, []
        for crua in r:
            linha = crua.decode('utf-8').rstrip('\r\n')
            if linha == '':
                if evento is not None:
                    try:
                        dados = json.loads(''.join(linhas)) if linhas else None
                    except ValueError:
                        dados = None
                    yield evento, dados
                evento, linhas = None, []
                continue
            if linha.startswith(':'):
                continue
            if linha.startswith('event:'):
                evento = linha[6:].strip()
            elif linha.startswith('data:'):
                linhas.append(linha[5:].strip())


def acompanhar(ident, ao_evento=None, env=None):
    """Segue o job até um estado final e o devolve. `ao_evento(evento, job)` a cada mudança.

    Um blip de rede entre o app e o portal não é o fim do job: o SSE reabre e continua do estado
    atual, que o portal manda ao conectar.
    """
    while True:
        final = None
        for evento, dados in _eventos('/jobs/%s/eventos' % ident, env=env):
            if ao_evento and dados is not None:
                try:
                    ao_evento(evento, dados)
                except Exception:
                    pass
            if evento in ('concluido', 'falhou', 'cancelado'):
                final = dados
        if final is not None:
            return final
        atual = estado(ident, env=env)
        if atual.get('estado') in ('concluido', 'falhou', 'cancelado'):
            return atual


def log(ident, env=None):
    """As linhas do container principal, como iterador; segue o pod enquanto ele vive."""
    for evento, dados in _eventos('/jobs/%s/log' % ident, env=env):
        if evento == 'linha' and dados:
            yield dados.get('linha', '')
        elif evento == 'error' and dados:
            raise ErroDaFila(0, dados.get('error') or 'erro no log')


def cancelar(ident, env=None):
    """Cancela; `True` quando o portal aceitou (um job já terminado também responde `True`)."""
    r = _pedir('POST', '/jobs/%s/cancelar' % ident, env=env)
    return bool(r and r.get('success'))


def remover(ident, env=None):
    """Apaga do S3 as entradas e as saídas de um job terminado, sem esperar a faxina da retenção.
    Serve a quem já baixou o que precisava e não quer o dado no cluster nem mais um minuto. Um job
    em curso responde 409: cancele antes."""
    r = _pedir('DELETE', '/jobs/%s' % ident, env=env)
    return bool(r and r.get('removido'))


def baixar(ident, destino, nomes=None, env=None):
    """Baixa as saídas para `destino` (criado se preciso) e devolve os caminhos gravados."""
    r = _pedir('GET', '/jobs/%s/saidas' % ident, env=env)
    os.makedirs(destino, exist_ok=True)
    gravados = []
    for s in r.get('saidas', []):
        if nomes is not None and s['nome'] not in nomes:
            continue
        caminho = os.path.join(destino, s['nome'])
        try:
            with urllib.request.urlopen(s['url'], timeout=None) as resp, open(caminho, 'wb') as f:
                while True:
                    b = resp.read(BLOCO)
                    if not b:
                        break
                    f.write(b)
        except urllib.error.HTTPError as e:
            raise ErroDaFila(e.code, 'o S3 recusou a saída %s (%d)' % (s['nome'], e.code))
        except urllib.error.URLError as e:
            raise ErroDaFila(0, 'o S3 não respondeu ao baixar %s: %s' % (s['nome'], e.reason))
        gravados.append(caminho)
    return gravados


def listar(env=None):
    """Os jobs deste app para este usuário, do mais novo ao mais velho."""
    r = _pedir('GET', '/jobs', env=env)
    return r.get('jobs', []) if r else []
